cuda/base/kernel_launch.cuh

```diff
@@ -0,0 +1,247 @@
+
+#ifndef GKO_CUDA_BASE_KERNEL_LAUNCH_CUH_
+#define GKO_CUDA_BASE_KERNEL_LAUNCH_CUH_
+
+
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+#include "cuda/base/device_guard.hpp"
+#include "cuda/base/types.hpp"
+#include "cuda/components/thread_ids.cuh"
+
+
+#ifdef GKO_KERNEL
+#error "Only one kernel_launch.hpp file can be included at a time."
+#else
+#define GKO_KERNEL __device__
+#endif
+
+
+namespace gko {
+namespace kernels {
+namespace cuda {
+
+
+constexpr auto default_block_size = 512;
+
+
+template <typename ValueType>
+struct matrix_accessor {
+    ValueType *data;
+    size_type stride;
+
+    __device__ ValueType &operator()(size_type row, size_type col)
+    {
+        return data[row * stride + col];
+    }
+
+    __device__ ValueType &operator[](size_type idx) { return data[idx]; }
+};
+
+
+template <typename T>
+struct to_device_type_impl {
+    using type = std::decay_t<cuda_type<T>>;
+    static type map_to_device(T in) { return as_cuda_type(in); }
+};
+
+template <typename ValueType>
+struct to_device_type_impl<matrix::Dense<ValueType> *&> {
+    using type = matrix_accessor<cuda_type<ValueType>>;
+    static type map_to_device(matrix::Dense<ValueType> *mtx)
+    {
+        return {as_cuda_type(mtx->get_values()), mtx->get_stride()};
+    }
+};
+
+template <typename ValueType>
+struct to_device_type_impl<const matrix::Dense<ValueType> *&> {
+    using type = matrix_accessor<const cuda_type<ValueType>>;
+    static type map_to_device(const matrix::Dense<ValueType> *mtx)
+    {
+        return {as_cuda_type(mtx->get_const_values()), mtx->get_stride()};
+    }
+};
+
+template <typename ValueType>
+struct to_device_type_impl<Array<ValueType> &> {
+    using type = cuda_type<ValueType> *;
+    static type map_to_device(Array<ValueType> &array)
+    {
+        return as_cuda_type(array.get_data());
+    }
+};
+
+template <typename ValueType>
+struct to_device_type_impl<const Array<ValueType> &> {
+    using type = const cuda_type<ValueType> *;
+    static type map_to_device(const Array<ValueType> &array)
+    {
+        return as_cuda_type(array.get_const_data());
+    }
+};
+
+
+template <typename T>
+typename to_device_type_impl<T>::type map_to_device(T &&param)
+{
+    return to_device_type_impl<T>::map_to_device(param);
+}
+
+
+template <typename ValueType>
+struct compact_dense_wrapper {
+    ValueType *data;
+};
+
+
+template <typename T>
+struct device_unpack_2d_impl {
+    using type = T;
+    static __device__ __forceinline__ type unpack(T param, size_type, size_type,
+                                                  size_type, size_type)
+    {
+        return param;
+    }
+};
+
+template <typename ValueType>
+struct device_unpack_2d_impl<compact_dense_wrapper<ValueType>> {
+    using type = matrix_accessor<ValueType>;
+    static __device__ __forceinline__ type
+    unpack(compact_dense_wrapper<ValueType> param, size_type, size_type,
+           size_type, size_type num_cols)
```

Comment:
Here, too, are these extra parameters necessary?
```suggestion
    unpack(compact_dense_wrapper<ValueType> param, size_type num_cols)
```

Reply:
At the moment no, they could only be useful for row-major storage at some point.

Reply:
Good point. But then you need just one more parameter for `num_rows`, because the lambda would be the current row and column indices anyway.
```suggestion
    unpack(compact_dense_wrapper<ValueType> param,
           size_type, size_type num_cols)
```